README.rst
pytracking - Email Open and Click Tracking Library
==================================================

:Authors:
  Resulto Developpement Web Inc.
:Version: 0.2.1.1

This library provides a set of functions that provide open and click tracking
when sending emails. This is particularly useful if you rely on an Email
Service Provider (ESP) which does not provide open and click tracking.

The library only provides building blocks and does not handle the actual
sending of email or the serving of tracking pixel and links, but it comes
pretty close to this.

.. image:: https://img.shields.io/pypi/v/pytracking.svg
    :target: https://pypi.python.org/pypi/pytracking

.. image:: https://img.shields.io/pypi/l/pytracking.svg

.. image:: https://img.shields.io/pypi/pyversions/pytracking.svg


.. contents:: Summary
    :backlinks: entry
    :local:


Overview
--------

There are two main steps when tracking email opens and link clicks:

1. Adding tracking information to emails
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To track email opens, the generally accepted strategy is to add a small 1x1
transparent pixel at the end of an email. When a user opens an email, the email
client (e.g., gmail, outlook, thunderbird) will load the pixel by making a GET
request. The web server serving the request will then record the open and
notify the sender of the email.

To track link clicks, the generally accepted strategy is to rewrite links in an
email to change the destination to a proxy. Once a user clicks on the link, the
proxy redirects the user to the real link and notifies the sender of the email.

pytracking provides a stateless strategy to open and click tracking: all the
information you want to track are encoded in the pixel (open) and proxy (click)
URLs. For example, if you want to track the customer id and the transaction id
associated with a particular email, pytracking will encode this information in
the URL. When the user opens the email or clicks on a link, the customer id and
transaction id will be decoded and can then be sent to a webhook.

See the `Get Open Tracking Link`_ section for a quick example.


2. Handling email opens and link clicks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Once a user opens an email or clicks on a link, the email client will send a
request to the encoded URL. Your web server will receive such request and pass
it to pytracking, which will decode the tracking information. You can then use
the tracking information directly (e.g., update your tracking database) or you
can send the information to a webhook.

In the case of link tracking, the decoded information will contain the original
URL that you must redirect the email client to.

See the `Get Open Tracking Data from URL`_ section for a quick example.



Optional Major Features provided by pytracking
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Encryption: pytracking uses base 64 to encode your tracking information,
   which can be decoded by anyone. You can optionaly encrypt your tracking
   information, which can only be decoded if you have the key. See the
   `Encrypting Data`_ section for more information.

2. HTML modification: pytracking can modify an HTML email to replace all links
   and add a tracking pixel. See the `Modifying HTML emails to add tracking
   links`_ section.

3. Django: if you use Django to serve open and click tracking URLs, you can
   extend pytracking Django views, which already provides the redirect and
   pixel serving. See the `Using pytracking with Django`_ section.

4. Webhooks: pytracking offers a shortcut function to make a POST request to a
   webhook. See the `Notifying Webhooks`_ section.


Requirements
------------

pytracking works with Python 3.6+. It doesn't require any external library, but
there are many optional features that have dependencies.


Installation
------------

You can install pytracking using pip:

::

    pip install pytracking

You can install specific features with extras:

::

    pip install pytracking[django,crypto]

You can also install all features:

::

    pip install pytracking[all]



Basic Library Usage
-------------------

You can generate two kinds of tracking links with pytracking: a link to a
transparent tracking pixel and a link that redirects to another link.

Encoding
~~~~~~~~

You can encode metadata in both kinds of links. For example, you can associate
a customer id with a click tracking link so when the customer clicks on the
link, you'll know exactly which customer clicked on it.

pylinktracking implements a stateless tracking strategy: all necessary
information can be encoded in the tracking links. You can optionally keep
common settings (e.g., default metadata to associate with all links, webhook
URL) in a separate configuration.

The information is encoded using url-safe base64 so anyone intercepting your
links, including your customers, could potentially decode the information. You
can optionally encrypt the tracking information (see below).

Most functions take as a parameter a ``pytracking.Configuration``
instance that tells how to generate the links. You can also pass the
configuration parameters as ``**kwargs`` argument or can mix both: the kwargs
will override the configuration parameters.

Decoding
~~~~~~~~

Once you get a request from a tracking link, you can use pytracking to decode
the link and get a ``pytracking.TrackingResult`` instance, which contains
information such as the link to redirect to (if it's a click tracking link),
the associated metadata, the webhook URL to notify, etc.

Basic Library Examples
----------------------

Get Open Tracking Link
~~~~~~~~~~~~~~~~~~~~~~

::

    import pytracking

    open_tracking_url = pytracking.get_open_tracking_url(
        {"customer_id": 1}, base_open_tracking_url="https://trackingdomain.com/path/",
        webhook_url="http://requestb.in/123", include_webhook_url=True)

    # This will produce a URL such as:
    # https://trackingdomain.com/path/e30203jhd9239754jh21387293jhf989sda=


Get Open Tracking Link with Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

::

    import pytracking

    configuration = pytracking.Configuration(
        base_open_tracking_url="https://trackingdomain.com/path/",
        webhook_url="http://requestb.in/123",
        include_webhook_url=False)

    open_tracking_url = pytracking.get_open_tracking_url(
        {"customer_id": 1}, configuration=configuration)

    # This will produce a URL such as:
    # https://trackingdomain.com/path/e30203jhd9239754jh21387293jhf989sda=


Get Click Tracking Link
~~~~~~~~~~~~~~~~~~~~~~~

::

    import pytracking

    click_tracking_url = pytracking.get_click_tracking_url(
        "http://www.example.com/?query=value", {"customer_id": 1},
        base_click_tracking_url="https://trackingdomain.com/path/",
        webhook_url="http://requestb.in/123", include_webhook_url=True)

    # This will produce a URL such as:
    # https://trackingdomain.com/path/e30203jhd9239754jh21387293jhf989sda=


Get Open Tracking Data from URL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

::

    import pytracking

    full_url = "https://trackingdomain.com/path/e30203jhd9239754jh21387293jhf989sda="
    tracking_result = pytracking.get_open_tracking_result(
        full_url, base_open_tracking_url="https://trackingdomain.com/path/")

    # Metadata is in tracking_result.metadata
    # Webhook URL is in tracking_result.webhook_url


Get Click Tracking Data from URL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

::

    import pytracking

    full_url = "https://trackingdomain.com/path/e30203jhd9239754jh21387293jhf989sda="
    tracking_result = pytracking.get_open_tracking_result(
        full_url, base_click_tracking_url="https://trackingdomain.com/path/")

    # Metadata is in tracking_result.metadata
    # Webhook URL is in tracking_result.webhook_url
    # Tracked URL to redirect to is in tracking_result.tracked_url


Get a 1x1 transparent PNG pixel
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

::

    import pytracking

    (pixel_byte_string, mime_type) = pytracking.get_open_tracking_pixel()



Encrypting Data
---------------

You can encrypt your encoded data to prevent third parties from accessing the
tracking data encoded in your link.

To use the encryption feature, you must install pytracking with
``pytracking[crypto]``, which uses the `cryptography Python library
<https://cryptography.io/en/latest/>`_.

Encrypting your data slightly increases the length of the generated URL.

::

    import pytracking
    from cryptography.fernet import Fernet

    key = Fernet.generate_key()

    # Encode
    click_tracking_url = pytracking.get_click_tracking_url(
        "http://www.example.com/?query=value", {"customer_id": 1},
        base_click_tracking_url="https://trackingdomain.com/path/",
        webhook_url="http://requestb.in/123", include_webhook_url=True,
        encryption_bytestring_key=key)

    # Decode
    tracking_result = pytracking.get_open_tracking_result(
        full_url, base_click_tracking_url="https://trackingdomain.com/path/",
        encryption_bytestring_key=key)


Using pytracking with Django
----------------------------

pytracking comes with View classes that you can extend and that handle open and
click tracking link request.

For example, the ``pytracking.django.OpenTrackingView`` will return a 1x1
transparent PNG pixel for GET requests. The
``pytracking.django.ClickTrackingView`` will return a 302 redirect response to
the tracked URL.

Both views will return a 404 response if the tracking URL is invalid. Both
views will capture the user agent and the user ip of the request. This
information will be available in TrackingResult.request_data.

You can extend both views to determine what to do with the tracking result
(e.g., call a webhook or submit a task to a celery queue). Finally, you can
encode your configuration parameters in your Django settings or you can compute
them in your view.

To use the django feature, you must install pytracking with
``pytracking[django]``.

Configuration parameters in Django settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can provide default configuration parameters in your Django settings by
adding this key in your settings file:

::

    PYTRACKING_CONFIGURATION = {
        "webhook_url": "http://requestb.in/123",
        "base_open_tracking_url": "http://tracking.domain.com/open/",
        "base_click_tracking_url": "http://tracking.domain.com/click/",
        "default_metadata": {"analytics_key": "123456"},
        "append_slash": True
    }


Extending default views
~~~~~~~~~~~~~~~~~~~~~~~

::

    from pytracking import Configuration
    from pytracking.django import OpenTrackingView, ClickTrackingView

    class MyOpenTrackingView(OpenTrackingView):

        def notify_tracking_event(self, tracking_result):
            # Override this method to do something with the tracking result.
            # tracking_result.request_data["user_agent"] and
            # tracking_result.request_data["user_ip"] contains the user agent
            # and ip of the client.
            send_tracking_result_to_queue(tracking_result)

        def notify_decoding_error(self, exception, request):
            # Called when the tracking link cannot be decoded
            # Override this to, for example, log the exception
            logger.log(exception)

        def get_configuration(self):
            # By defaut, fetchs the configuration parameters from the Django
            # settings. You can return your own Configuration object here if
            # you do not want to use Django settings.
            return Configuration()


    class MyClickTrackingView(ClickTrackingView):

        def notify_tracking_event(self, tracking_result):
            # Override this method to do something with the tracking result.
            # tracking_result.request_data["user_agent"] and
            # tracking_result.request_data["user_ip"] contains the user agent
            # and ip of the client.
            send_tracking_result_to_queue(tracking_result)

        def notify_decoding_error(self, exception, request):
            # Called when the tracking link cannot be decoded
            # Override this to, for example, log the exception
            logger.log(exception)

        def get_configuration(self):
            # By defaut, fetchs the configuration parameters from the Django
            # settings. You can return your own Configuration object here if
            # you do not want to use Django settings.
            return Configuration()

URLs configuration
~~~~~~~~~~~~~~~~~~

Add this to your urls.py file:

::

    urlpatterns = [
        url(
            "^open/(?P<path>[\w=-]+)/$", MyOpenTrackingView.as_view(),
            name="open_tracking"),
        url(
            "^click/(?P<path>[\w=-]+)/$", MyClickTrackingView.as_view(),
            name="click_tracking"),
    ]


Notifying Webhooks
------------------

You can send a POST request to a webhook with the tracking result. The webhook
feature just packages the tracking result as a json string in the POST body. It
also sets the content encoding to ``application/json``.

To use the webhook feature, you must install pytracking with
``pytracking[webhook]``.


::

    import pytracking
    from pytracking.webhook import send_webhook

    # Assumes that the webhook url is encoded in the url.
    full_url = "https://trackingdomain.com/path/e30203jhd9239754jh21387293jhf989sda="
    tracking_result = pytracking.get_open_tracking_result(
        full_url, base_click_tracking_url="https://trackingdomain.com/path/")

    # Will send a POST request with the following json str body:
    #  {
    #    "is_open_tracking": False,
    #    "is_click_tracking": True,
    #    "metadata": {...},
    #    "request_data": None,
    #    "tracked_url": "http://...",
    #    "timestamp": 1389177318
    #  }
    send_webhook(tracking_result)



Modifying HTML emails to add tracking links
-------------------------------------------

If you have an HTML email, pytracking can update all links with tracking links
and it can also add a transparent tracking pixel at the end of your email.

To use the HTML feature, you must install pytracking with ``pytracking[html]``,
which uses the `lxml library <http://lxml.de/>`_.

::

    import pytracking
    from pytracking.html import adapt_html

    html_email_text = "..."
    new_html_email_text = adapt_html(
        html_email_text, extra_metadata={"customer_id": 1},
        click_tracking=True, open_tracking=True)


Testing pytracking
------------------

pytracking uses `tox <https://tox.readthedocs.io/en/latest/>`_ and `py.test
<http://docs.pytest.org/en/latest/>`_. If you have tox installed, just run
``tox`` and all possible configurations of pytracking will be tested on Python
3.6.


TODO
----

1. Add various checks to ensure that the input data is sane and does not bust
   any known limits (e.g., URL length).

2. Add more examples.

3. Allow multiplpe webhooks and webhooks per tracking method.

4. Transform Django views into view mixins.

5. Add option to encode the webhook timeout in the tracking URL.

6. Document caveats of using pytracking.html (example: long emails are often
   cut off by the email clients and the tracking pixel is thus not loaded).

7. Add some form of API documentation (at least Configuration and
   TrackingResult), maybe as a separate document.

License
-------

This software is licensed under the `New BSD License`. See the `LICENSE` file
in the repository for the full license text.


CHANGELOG.rst
Changelog - pytracking
======================

0.2.1.1 - November 24th 2022
----------------------------

- Fixed missing DOCTYPE and meta charset tag when adapting html for open
  and click tracking.

0.2.1 - October 28th 2020
-------------------------

- Added append_slash configuration parameter to append a slash to click and
  open tracking URLs
- Updated dependencies to prevent breaking changes in django-ipware
- Updated supported Python version to 3.6
- Updated README file
- Removed CircleCI integration

0.2.0 - June 6th 2017
---------------------

- Added request parameter to notify_decoding_error callback in
  ``pytracking.django.TrackingView``
- Added a reference to professional services in README.rst


0.1.0 - August 10th 2016
------------------------

- First release!
- Support encoding and decoding of open/click tracking links.
- Basic Django Views.
- Can send webhook POST requests with tracking data.
- Can encrypt tracking data instead of relying on base64.
- Can modify HTML content to replace links and add tracking pixel.


test_html.py
import pytest

from pytracking import (
    get_click_tracking_url_path, get_open_tracking_url_path,
    get_open_tracking_result, get_click_tracking_result)

from test_pytracking import (
    DEFAULT_BASE_OPEN_TRACKING_URL, DEFAULT_BASE_CLICK_TRACKING_URL,
    DEFAULT_METADATA, DEFAULT_WEBHOOK_URL, DEFAULT_DEFAULT_METADATA,
    EXPECTED_METADATA)

TEST_HTML_EMAIL = """
<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8"></meta>
<head>
</head>
<body>
<h1>Hello World Some éàù and &eacute; entities.</h1>
<p>
<a href="http://www.example.com/foo/?question=response">This is an inline link
</a> with some text.
</p>
<p>
<a href="mailto:bob@bob.com">Mail to Bob!</a>
</p>
<ul>
<li><a href="http://www.domain2.com">
<img src="http://www.test.com/test.jpg"></a></li>
</ul>
</body>
</html>
"""


DEFAULT_SETTINGS = {
    "webhook_url": DEFAULT_WEBHOOK_URL,
    "base_open_tracking_url": DEFAULT_BASE_OPEN_TRACKING_URL,
    "base_click_tracking_url": DEFAULT_BASE_CLICK_TRACKING_URL,
    "default_metadata": DEFAULT_DEFAULT_METADATA
}

try:
    from lxml import html  # noqa

    import pytracking.html as tracking_html

    support_html = True
except ImportError:
    support_html = False


pytestmark = pytest.mark.skipif(
    not support_html, reason="HTML-support lib not installed")


def test_adapt_html_full():
    new_html = tracking_html.adapt_html(
        TEST_HTML_EMAIL, DEFAULT_METADATA, **DEFAULT_SETTINGS)

    tree = html.fromstring(new_html)

    _test_open_tracking(tree)
    _test_click_tracking(tree)

    assert "<!DOCTYPE html>" in new_html
    assert "<meta charset=\"UTF-8\">" in new_html


def test_adapt_html_click_only():
    new_html = tracking_html.adapt_html(
        TEST_HTML_EMAIL, DEFAULT_METADATA, open_tracking=False,
        **DEFAULT_SETTINGS)

    tree = html.fromstring(new_html)
    assert len(tree.xpath("//img")) == 1

    _test_click_tracking(tree)


def test_adapt_html_open_only():
    new_html = tracking_html.adapt_html(
        TEST_HTML_EMAIL, DEFAULT_METADATA, click_tracking=False,
        **DEFAULT_SETTINGS)

    tree = html.fromstring(new_html)

    _test_open_tracking(tree)

    links = tree.xpath("//a")

    assert links[0].attrib["href"] ==\
        "http://www.example.com/foo/?question=response"
    assert links[1].attrib["href"] == "mailto:bob@bob.com"
    assert links[2].attrib["href"] == "http://www.domain2.com"


def _test_click_tracking(tree):
    links = tree.xpath("//a")
    first_link_url_path = get_click_tracking_url_path(
        links[0].attrib["href"], **DEFAULT_SETTINGS)
    click_result = get_click_tracking_result(
        first_link_url_path, **DEFAULT_SETTINGS)
    assert click_result.metadata == EXPECTED_METADATA
    assert click_result.tracked_url ==\
        "http://www.example.com/foo/?question=response"

    assert links[1].attrib["href"] == "mailto:bob@bob.com"

    second_link_url_path = get_click_tracking_url_path(
        links[2].attrib["href"], **DEFAULT_SETTINGS)
    click_result = get_click_tracking_result(
        second_link_url_path, **DEFAULT_SETTINGS)
    assert click_result.metadata == EXPECTED_METADATA
    assert click_result.tracked_url == "http://www.domain2.com"


def _test_open_tracking(tree):
    pixel_img = tree.xpath("//img")[-1]
    open_url_path = get_open_tracking_url_path(
        pixel_img.attrib["src"], **DEFAULT_SETTINGS)
    open_result = get_open_tracking_result(
        open_url_path, **DEFAULT_SETTINGS)

    assert open_result.metadata == EXPECTED_METADATA


test_crypto.py
import copy

import pytest

from pytracking import (
    Configuration, get_click_tracking_url, get_click_tracking_url_path,
    get_open_tracking_url_path,
    get_click_tracking_result, get_open_tracking_url,
    get_open_tracking_result)


try:
    from cryptography.fernet import Fernet
    support_crypto = True
except ImportError:
    support_crypto = False

pytestmark = pytest.mark.skipif(
    not support_crypto, reason="Cryptography lib not installed")

DEFAULT_URL_TO_TRACK = "https://www.bob.com/hello-world/?token=valueééé"

DEFAULT_BASE_CLICK_TRACKING_URL = "https://a.b.com/tracking/"

DEFAULT_BASE_OPEN_TRACKING_URL = "https://a.b.com/tracking/open/"

DEFAULT_ENCRYPTION_KEY = b'XdhWbQZnqCIPLBL0ViPIW2vBTsmUNxAS-7mOtTdu6ZM='

DEFAULT_METADATA = {
    "param1": "val1",
    "param3": "val3b",
    "nested": {"param2": "val2"}}

DEFAULT_DEFAULT_METADATA = {
    "key1": True,
    "keyéé": "valèèè",
    "param3": "val3"
}

DEFAULT_WEBHOOK_URL = "https://webhook.com/tracking/"

DEFAULT_REQUEST_DATA = {
    "user_agent": "Firefox",
    "user_ip": "127.0.0.1"
}

DEFAULT_CONFIGURATION = Configuration(
    webhook_url=DEFAULT_WEBHOOK_URL,
    base_open_tracking_url=DEFAULT_BASE_OPEN_TRACKING_URL,
    base_click_tracking_url=DEFAULT_BASE_CLICK_TRACKING_URL,
    default_metadata=DEFAULT_DEFAULT_METADATA)


def test_basic_encrypted_get_open_tracking_url():
    url = get_open_tracking_url(
        base_open_tracking_url=DEFAULT_BASE_OPEN_TRACKING_URL,
        encryption_bytestring_key=DEFAULT_ENCRYPTION_KEY)
    path = get_open_tracking_url_path(
        url, base_open_tracking_url=DEFAULT_BASE_OPEN_TRACKING_URL)
    key = Fernet(DEFAULT_ENCRYPTION_KEY)

    # Can decrypt without raising an exception
    value = key.decrypt(path.encode("utf-8"))
    # We can only assert if the value is truthy because the value is encrypted
    # with a different salt each time.
    assert value


def test_minimal_encrypted_get_open_tracking_url():
    url = get_open_tracking_url(
        base_open_tracking_url=DEFAULT_BASE_OPEN_TRACKING_URL,
        metadata=DEFAULT_METADATA,
        encryption_bytestring_key=DEFAULT_ENCRYPTION_KEY)
    path = get_open_tracking_url_path(
        url, base_open_tracking_url=DEFAULT_BASE_OPEN_TRACKING_URL)

    tracking_result = get_open_tracking_result(
        path,
        request_data=DEFAULT_REQUEST_DATA,
        default_metadata=DEFAULT_DEFAULT_METADATA,
        webhook_url=DEFAULT_WEBHOOK_URL,
        encryption_bytestring_key=DEFAULT_ENCRYPTION_KEY)

    expected_metadata = copy.copy(DEFAULT_DEFAULT_METADATA)
    expected_metadata.update(DEFAULT_METADATA)

    assert tracking_result.tracked_url is None
    assert tracking_result.webhook_url == DEFAULT_WEBHOOK_URL
    assert tracking_result.request_data == DEFAULT_REQUEST_DATA
    assert tracking_result.metadata == expected_metadata
    assert tracking_result.is_open_tracking
    assert not tracking_result.is_click_tracking


def test_basic_encrypted_get_click_tracking_url():
    url = get_click_tracking_url(
        DEFAULT_URL_TO_TRACK,
        base_click_tracking_url=DEFAULT_BASE_CLICK_TRACKING_URL,
        encryption_bytestring_key=DEFAULT_ENCRYPTION_KEY)
    path = get_click_tracking_url_path(
        url, base_click_tracking_url=DEFAULT_BASE_CLICK_TRACKING_URL)
    key = Fernet(DEFAULT_ENCRYPTION_KEY)

    # Can decrypt without raising an exception
    value = key.decrypt(path.encode("utf-8"))
    # We can only assert if the value is truthy because the value is encrypted
    # with a different salt each time.
    assert value


def test_minimal_encrypted_get_click_tracking_url():
    url = get_click_tracking_url(
        DEFAULT_URL_TO_TRACK,
        base_click_tracking_url=DEFAULT_BASE_CLICK_TRACKING_URL,
        metadata=DEFAULT_METADATA,
        encryption_bytestring_key=DEFAULT_ENCRYPTION_KEY)
    path = get_click_tracking_url_path(
        url, base_click_tracking_url=DEFAULT_BASE_CLICK_TRACKING_URL)

    tracking_result = get_click_tracking_result(
        path,
        request_data=DEFAULT_REQUEST_DATA,
        default_metadata=DEFAULT_DEFAULT_METADATA,
        webhook_url=DEFAULT_WEBHOOK_URL,
        encryption_bytestring_key=DEFAULT_ENCRYPTION_KEY)

    expected_metadata = copy.copy(DEFAULT_DEFAULT_METADATA)
    expected_metadata.update(DEFAULT_METADATA)

    assert tracking_result.tracked_url == DEFAULT_URL_TO_TRACK
    assert tracking_result.webhook_url == DEFAULT_WEBHOOK_URL
    assert tracking_result.request_data == DEFAULT_REQUEST_DATA
    assert tracking_result.metadata == expected_metadata
    assert tracking_result.is_click_tracking
    assert not tracking_result.is_open_tracking


pytracking/html.py
from lxml import html

from pytracking.tracking import (
    get_configuration, get_open_tracking_url, get_click_tracking_url)


DEFAULT_ATTRIBUTES = {
    "border": "0",
    "width": "0",
    "height": "0",
    "alt": ""
}

DOCTYPE = "<!DOCTYPE html>"


def adapt_html(
        html_text, extra_metadata, click_tracking=True, open_tracking=True,
        configuration=None, **kwargs):
    """Changes an HTML string by replacing links (<a href...>) with tracking
    links and by adding a 1x1 transparent pixel just before the closing body
    tag.

    :param html_text: The HTML to change (unicode or bytestring).
    :param extra_metadata: A dict that can be json-encoded and that will
        be encoded in the tracking link.
    :param click_tracking: If links (<a href...>) must be changed.
    :param open_tracking: If a transparent pixel must be added before the
        closing body tag.
    :param configuration: An optional Configuration instance.
    :param kwargs: Optional configuration parameters. If provided with a
        Configuration instance, the kwargs parameters will override the
        Configuration parameters.
    """
    configuration = get_configuration(configuration, kwargs)

    tree = html.fromstring(html_text)

    if click_tracking:
        _replace_links(tree, extra_metadata, configuration)

    if open_tracking:
        _add_tracking_pixel(tree, extra_metadata, configuration)

    new_html_text = html.tostring(
        tree, include_meta_content_type=True, doctype=DOCTYPE)

    return new_html_text.decode("utf-8")


def _replace_links(tree, extra_metadata, configuration):
    for (element, attribute, link, pos) in tree.iterlinks():
        if element.tag == "a" and attribute == "href" and _valid_link(link):
            new_link = get_click_tracking_url(
                link, extra_metadata, configuration)
            element.attrib["href"] = new_link


def _add_tracking_pixel(tree, extra_metadata, configuration):
    url = get_open_tracking_url(extra_metadata, configuration)
    pixel = html.Element("img", {"src": url})
    tree.body.append(pixel)


def _valid_link(link):
    return link.startswith("http://") or link.startswith("https://") or\
        link.startswith("//")


pytracking/webhook.py
import requests

from pytracking.tracking import get_configuration


def send_webhook(tracking_result, configuration=None, **kwargs):
    """Sends a POST request to the webhook URL specified in tracking_result.

    The POST request will have a body of type application/json that contains a
    json representation of the tracking result:

    ::

        {
            "is_open_tracking": False,
            "is_click_tracking": True,
            "metadata": {...},
            "request_data": None,
            "tracked_url": "http://...",
            "timestamp": 1389177318
        }

    :param tracking_result: The TrackingResult instance to post to a webhook.
    :param configuration: An optional Configuration instance.
    :param kwargs: Optional configuration parameters. If provided with a
        Configuration instance, the kwargs parameters will override the
        Configuration parameters.
    """
    configuration = get_configuration(configuration, kwargs)

    payload = {
        "is_open_tracking": tracking_result.is_open_tracking,
        "is_click_tracking": tracking_result.is_click_tracking,
        "metadata": tracking_result.metadata,
        "request_data": tracking_result.request_data,
        "timestamp": tracking_result.timestamp
    }

    if tracking_result.tracked_url:
        payload["tracked_url"] = tracking_result.tracked_url

    response = requests.post(
        tracking_result.webhook_url, json=payload,
        timeout=configuration.webhook_timeout_seconds)

    return response


pytracking/django.py
from django.conf import settings
from django.http import (
    HttpResponseRedirect, Http404, HttpResponse)
from django.views.generic import View
from ipware import get_client_ip
from pytracking.tracking import (
    get_configuration, TRACKING_PIXEL, PNG_MIME_TYPE)


class TrackingView(View):
    """Base Tracking View.

    Subclasses should override notify_* methods.
    """

    def notify_tracking_event(self, tracking_result):
        """Called once the tracking link has been decoded, and before
        responding with a redirect or a tracking pixel.

        :param tracking_result: An instance of TrackingResult.
        """
        pass

    def notify_decoding_error(self, exception, request):
        """Called when a decoding error occurs, and before
        responding with a 404.

        :param exception: The exception that was raised when trying to decode a
            tracking link.
        """
        pass

    def get_configuration(self):
        """Returns a Configuration instance built from
        settings.PYTRACKING_CONFIGURATION.

        Override this method if you want to build your own Configuration
        instance.
        """
        return get_configuration_from_settings()


class ClickTrackingView(TrackingView):
    """View that decodes a URL, call notify_tracking_event, and then sends a
    302 redirect to the tracked URL.

    This view captures the user agent and the ip of the client that makes a
    request and puts this information in the request_data dictionary under the
    ``user_agent`` and ``user_ip`` keys.

    If a decoding error occurs, notify_decoding_error is called and a 404 is
    returned.

    If no tracking url is present in the decoded URL, a 404 is returned.
    """

    def get(self, request, path):
        configuration = self.get_configuration()

        try:
            tracking_result = get_tracking_result(
                request, path, False, configuration)
        except Exception as e:
            self.notify_decoding_error(e, request)
            raise Http404

        if not tracking_result.tracked_url:
            raise Http404

        self.notify_tracking_event(tracking_result)

        return HttpResponseRedirect(tracking_result.tracked_url)


class OpenTrackingView(TrackingView):
    """View that decodes a URL, call notify_tracking_event, and then returns a
    1x1 transparent PNG file with a image/png mime type.

    This view captures the user agent and the ip of the client that makes a
    request and puts this information in the request_data dictionary under the
    ``user_agent`` and ``user_ip`` keys.

    If a decoding error occurs, notify_decoding_error is called and a 404 is
    returned.
    """

    def get(self, request, path):
        configuration = self.get_configuration()

        try:
            tracking_result = get_tracking_result(
                request, path, True, configuration)
        except Exception as e:
            self.notify_decoding_error(e, request)
            raise Http404

        self.notify_tracking_event(tracking_result)

        return HttpResponse(TRACKING_PIXEL, content_type=PNG_MIME_TYPE)


def get_request_data(request):
    """Retrieves the user agent and the ip of the client from the Django
    request.

    We use ipware to retrieve the "real" IP (e.g., load balancer will often put
    the client IP in X-Forwarded-For header).
    """
    user_agent = request.META.get("HTTP_USER_AGENT")
    ip = get_client_ip(request)[0]
    return {
        "user_agent": user_agent,
        "user_ip": ip
    }


def get_configuration_from_settings(settings_name="PYTRACKING_CONFIGURATION"):
    """Builds a Configuration instance from the parameters in
    settings.PYTRACKING_CONFIGURATION.
    """
    kwargs = getattr(settings, settings_name)
    return get_configuration(None, kwargs)


def get_tracking_result(request, path, is_open, configuration=None, **kwargs):
    """Builds a tracking result from a Django request.

    :param request: A Django request
    :param path: The path of the URL that is encoded (usually possible to
        distinguish this path using a URL parameter).
    :param is_open: If this is an open tracking link request.
    :param configuration: An optional Configuration instance.
    :param kwargs: Optional configuration parameters. If provided with a
        Configuration instance, the kwargs parameters will override the
        Configuration parameters.
    """
    configuration = get_configuration(configuration, kwargs)
    request_data = get_request_data(request)
    return configuration.get_tracking_result(
        path, request_data, is_open)
